src/DataStructures/Mesh.hpp

```diff
@@ -0,0 +1,177 @@
+   * number of collocation points.
+   *
+   * \param extents The number of collocation points per dimension
+   * \param basis The choice of spectral basis to compute the
+   * collocation points
+   * \param quadrature The choice of quadrature to compute
+   * the collocation points
+   */
+  Mesh(std::array<size_t, Dim> extents, const Spectral::Basis basis,
+       const Spectral::Quadrature quadrature) noexcept
+      : extents_(std::move(extents)) {
+    bases_.fill(basis);
+    quadratures_.fill(quadrature);
+  }
+
+  /*!
+   * Construct a computational grid where each dimension can have both a
+   * different number and placement of collocation points.
+   *
+   * \param extents The number of collocation points per dimension
+   * \param bases The choice of spectral bases to compute the
+   * collocation points per dimension
+   * \param quadratures The choice of quadratures to compute
+   * the collocation points per dimension
+   */
+  Mesh(std::array<size_t, Dim> extents, std::array<Spectral::Basis, Dim> bases,
+       std::array<Spectral::Quadrature, Dim> quadratures) noexcept
+      : extents_(std::move(extents)),
+        bases_(std::move(bases)),
+        quadratures_(std::move(quadratures)) {}
+
+  /*!
+   * \brief The number of grid points in each dimension of the grid.
+   */
+  const Index<Dim>& extents() const noexcept { return extents_; }
+
+  /*!
+   * \brief The number of grid points in dimension `d` of the grid.
+   */
+  size_t extents(const size_t d) const noexcept { return extents_[d]; }
+
+  /*!
+   * \brief The total number of grid points in all dimensions.
+   *
+   * \details `DataVector`s that represent field values on the grid have this
+   * many entries.
+   *
+   * \note A zero-dimensional mesh has one grid point, since it is the slice
+   * through a one-dimensional mesh (a line).
+   */
+  size_t number_of_grid_points() const noexcept { return extents_.product(); }
+
+  /*!
+   * \brief The basis chosen in each dimension of the grid.
+   */
+  const std::array<Spectral::Basis, Dim>& basis() const noexcept {
+    return bases_;
+  }
+
+  /*!
+   * \brief The basis chosen in dimension `d` of the grid.
+   */
+  Spectral::Basis basis(const size_t d) const noexcept {
+    return gsl::at(bases_, d);
+  }
+
+  /*!
+   * \brief The quadrature chosen in each dimension of the grid.
+   */
+  const std::array<Spectral::Quadrature, Dim>& quadrature() const noexcept {
+    return quadratures_;
+  }
+
+  /*!
+   * \brief The quadrature chosen in dimension `d` of the grid.
+   */
+  Spectral::Quadrature quadrature(const size_t d) const noexcept {
+    return gsl::at(quadratures_, d);
+  }
+
+  /*!
+   * \brief Returns a mesh with dimension `d` removed.
+   *
+   * \param d The dimension to remove (zero-indexed).
+   */
+  // clang-tidy: incorrectly reported redundancy in template expression
+  template <size_t N = Dim, Requires<(N > 0 and N == Dim)> = nullptr>  // NOLINT
```

Comment:
`clang-tidy` tells me this condition is redundant. Could someone have a look at this and confirm? We just want to remove the `slice_away` method from a zero-dimensional mesh here.

Reply:
Yes, that would be a clang-tidy bug. It is redundant is some cases, but not all, and it is the ones where it is not redundant we care about :) In short, I'm fine with NOLINT-ing this